src/libraries/System.IO.Pipes/src/System/IO/Pipes/NamedPipeClientStream.cs

```diff
@@ -135,7 +135,7 @@ private void ConnectInternal(int timeout, CancellationToken cancellationToken, i
                 cancellationToken.ThrowIfCancellationRequested();
 
                 // Determine how long we should wait in this connection attempt
-                int waitTime = timeout - elapsed;
+                int waitTime = timeout == Timeout.Infinite ? CancellationCheckInterval : timeout - elapsed;
```

Comment:
I can see that the next line was supposed to prevent this issue:

```cs
if (cancellationToken.CanBeCanceled && waitTime > CancellationCheckInterval)
```

the problem is that `waitTime` is `-1` here and `waitTime > CancellationCheckInterval` is false.

Your fix is correct when `cancellationToken.CanBeCanceled` is true, but not when it's false.

```suggestion
                int waitTime = timeout == Timeout.Infinite  && cancellationToken.CanBeCanceled? CancellationCheckInterval : timeout - elapsed;
```

Reply:
I think in case of the infinite timeout we should always use the `CancellationCheckInterval`.

1. If the token can be cancelled we will check it state.

2. If the token can't be cancelled we can pass `-1` further to the `TryConnect` and wait the method forever. But it breaks an abstraction of the `TryConnect` because the timeout parameter of the method is `int` without any documentation for any specific values like the infinite timeout. I think the best way is to assume that the timeout parameter of the `TryConnect` accepts simple positive timeout values and use `CancellationCheckInterval` instead.

When the timeout is finite and the token can be cancelled we should check it every `CancellationCheckInterval`. But if the token can't be cancelled we can pass the timeout as is to the `TryConnect` and just return after the all timeout exceeded.

As I see the implemented code fits that explained logic well.

Here is a table with all possible results depending on input.

| `timeout` | `waitTime:138`               |  `cancellationToken.CanBeCancelled` | `waitTime > CancellationCheckInterval` | Final Wait Time               |
|-----------|------------------------------|-------------------------------------|----------------------------------------|-------------------------------|
| Infinite  | `CancellationCheckInterval`  |  `true`                             | `false`                                | `CancellationCheckInterval`   |
| Finite    | `timeout - elapsed`          |  `true`                             | `true`                                 |  `CancellationCheckInterval`  |
| Infinite  |  `CancellationCheckInterval` |  `false`                            | `false`                                |   `CancellationCheckInterval` |
| Finite    |  `timeout - elapsed`         |  `false`                            | `true`                                 |   `timeout - elapsed`  which resolves to just single use `timeout`       |
|  Finite   |  `timeout - elapsed`         |   `true`                            |  `false`                               |  `timeout - elapsed`  and less than `CancellationCheckInterval`       |
|  Finite   |   `timeout - elapsed`        |   `false`                           |  `false`                               |   `timeout - elapsed`  and less than `CancellationCheckInterval`       |

Reply:
After thinking about it for a while I believe it's good to always use `CancellationCheckInterval` for `Timeout.Infinite` as it provides more optimal resource consumption (the thread is not blocked, it might yield and perform some other work).